runtime-rs: handle copy files when share_fs is not available

ananos opened the pull request:
In hypervisors that do not support virtiofs we have to copy files in the VM sandbox to properly setup the network (resolv.conf, hosts, and hostname).

To do that, we construct the volume as before, with the addition of an extra variable that designates the path where the file will reside in the sandbox.

In this case, we issue a `copy_file` agent request *and* we patch the spec to account for this change.

Fixes: #6978

Tim-0731-Hzt commented on src/runtime-rs/crates/resource/src/volume/share_fs_volume.rs:
```diff
@@ -60,10 +63,21 @@ impl ShareFsVolume {
+                // This is where we set the value for the guest path
+                volume.guest_path = dest;
                 if src.is_file() {
-                    // TODO: copy file
```
I think it's better to implement copy file here. 
1. if `share_fs` match None, it means the hypervisor doesn't support fs_share, so we don't need to check  `is_fs_sharing_supported` again in `create`. 
3. we could simplify the `create` function, and the code will look more readable

Tim-0731-Hzt commented on src/runtime-rs/crates/resource/src/volume/share_fs_volume.rs:
```diff
@@ -231,6 +245,17 @@ impl Volume for ShareFsVolume {
         Ok(None)
     }
+
+    fn get_guest_path(&self) -> Result<Option<String>> {
```
I think you don't need this trait, if copy file is implemented in this file, the `guest path` will be stored in mount, `get_volume_mount` will be enough for use.

Tim-0731-Hzt reviewed (commented):
Thanks @ananos few comments

fidencio commented:
Sorry, at the time you opened your PR we were having issues with our CI, which leads to required tests failing without having nothing to do with this PR.

AS this has been already fixed on  main, we´d like to ask you to rebase your PR atop of the current `main` branch.

Thanks for the consideration, and sorry for the trouble.

ananos commented:
thanks @Tim-0731-Hzt for the comments! really appreciate your prompt response.

We weren't sure how to handle the copy request to the agent in share_fs_volume, that's why we did this trampoline thing with manager/manager_inner and container.rs. It seems that passing the agent as a parameter to the constructor of the share_fs_volume resolves this.

Let us know if this latest commit addresses your concerns. Thanks again!

Tim-0731-Hzt commented on src/runtime-rs/crates/resource/src/volume/share_fs_volume.rs:
```diff
@@ -60,10 +64,51 @@ impl ShareFsVolume {
+                    .context("copy file request failed")?;
+
+                // This is where we set the value for the guest path
                 if src.is_file() {
```
Thanks @ananos , this PR generally LGTM.  One more question, should the operation above be handled in this if statement?  If it is a none regular file, I think we don't need to copy this guest ?

ananos replied:
you're absolutely right! our bad -- should be now fixed.

ananos commented:
https://github.com/kata-containers/kata-containers/actions/runs/5134329580/jobs/9238180586?pr=6980#step:10:1914 <- could this be related to this: https://github.com/rust-lang/rust/issues/63033 ?

do we mind if we move to rust 1.69.0?

Tim-0731-Hzt commented:
> https://github.com/kata-containers/kata-containers/actions/runs/5134329580/jobs/9238180586?pr=6980#step:10:1914 <- could this be related to this: [rust-lang/rust#63033](https://github.com/rust-lang/rust/issues/63033) ?
> 
> do we mind if we move to rust 1.69.0?



It still could work without upgrading to 1.69.0. if you pass the Agent instance as a reference, just like this
```rust
pub(crate) async fn new(
      share_fs: &Option<Arc<dyn ShareFs>>,
      m: &oci::Mount,
      cid: &str,
      readonly: bool,
      agent: &dyn Agent,
  )
```
But I think it's better to use `Arc<dyn Agent>`, as it allows the Agent instance could be shared, and be dropped if no more reference to it. Maybe we should consider move to 1.69 ? cc @quanweiZhou

Ji-Xinyou commented:
/test

ananos commented:
/test

ananos commented:
/retest-power

fidencio commented:
@ananos, @Tim-0731-Hzt, not exactly related to this PR, but I'd love to see a `SharedFS: none` option in the runtime-rs, which would explicitly copy the files to the guest in case users only want to use devmapper without having to pay the penalty of still having virtiofsd / nydus processes around.

fidencio reviewed (approved):
lgtm, thanks @ananos!
